### PR TITLE
Initial feature [WIP]

### PR DESCRIPTION
**Pull request by AdiJoost**
This branch is still work in progress.

Missing: Spawning Agents with a target, additional translocomotion, unit tests

### CODE REVIEW

File: src/simulation/locomotionAlgorithms/dijkstra.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        for newX in range(currentNode[0] - 1, currentNode[0] + 2):
+            for newY in range(currentNode[1] - 1, currentNode[1] + 2):
+                if self.grid.contains(newX, newY) and self.grid.getTileStatus(newX, newY) != TileStatus.BLOCKED:
+                    distance = 1 if self._isDiagonalStep(currentNode[0], currentNode[1], newX, newY) else np.sqrt(2)
+                    newValue = currentNode[2] + distance
+                    if (self.pedestrianHeatMap.getValue(newX, newY) > newValue):
+                        self.pedestrianHeatMap.updateValue(newX, newY, newValue)
+                        self.stack.append((newX, newY, newValue))
+    
+    def _isDiagonalStep(self, x:int, y:int, xNew:int, yNew: int) -> bool:
```
Comment (DrGeli) [inline src/simulation/locomotionAlgorithms/dijkstra.py:47]: sure this method does what it says?

File: src/simulation/spawner/baseSpawner.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            if (self.spawnerTiles[currentSpawnerTilePosition].tileDTO.getTileStatus() == TileStatus.FREE):
+                self.spawnerTiles[currentSpawnerTilePosition].changeState(TileStatus.PEDESTRIAN)
+                spawned += 1
+            currentSpawnerTilePosition += 1
+
+
+    def _getNumberOfSpawns(self) -> float:
+        return min(self.maxSpawnsPerBatch, self.numberOfTotalSpawns - self.totalSpawend, self._getNumberOfFreeTiles())
+
+    def _getNumberOfFreeTiles(self) -> float:
```
Comment (DrGeli) [inline src/simulation/spawner/baseSpawner.py:47]: why float?

File: src/exceptions/simulationErrorCodes.py [hunk 3]
```diff
@@ -8,6 +8,9 @@ class ErrorCodeDataMixin:
 class SimulationErrorCodes(ErrorCodeDataMixin, Enum):
     ExampleError = 1, "Example ErrorCode"
     NOT_IMPLEMENTED_IN_SIMULATION = 2, "A Function in the simulation is not implemented"
+    LENGTH_OF_GRID_INVALID = 3, "The legth of the grid has to be more than 0 and of type int"
```
Comment (DrGeli) [inline src/exceptions/simulationErrorCodes.py:11]: typo

File: src/enums/locomotionAlgorithms.py [hunk 4]
```diff
@@ -0,0 +1,4 @@
+from enum import Enum
+
+class LocomationAlgorihms(Enum):
```
Comment (DrGeli) [inline src/enums/locomotionAlgorithms.py:3]: typo

File: src/config/config.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        self.maxItterations = self._DEFAULT_MAX_NUMBER_OF_ITTERATIONS
+        pass
+
+    def getSavePath(self) -> str:
+        return self.savePath
+
+    def setSavePath(self, path: str) -> None:
+        self.savePath = path
+
+    def getMaxItterations(self) -> int:
```
Comment (DrGeli) [inline src/config/config.py:18]: typo

File: src/dtoModule/tileValueDTO.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
+
+
+class TileValueDTO():
```
Comment (DrGeli) [inline src/dtoModule/tileValueDTO.py:3]: what does DTO stand for? Please do not use abbreviations that are not clear and write it out.

**Review by DrGeli — COMMENTED** [review]
Dear Group B,

the code looks very well structured so far and easy to understand. 

This is what I recognized while reviewing:
- There are several typos -> please pay attention for correct spellings, although it does not change behaviour and thus might seem to you not so important, it reduces readability (and navigation through code) if classes /methods are misspelled
- There is a lot of structure for error handling etc. but not much functionality so far. Please be careful that you don't run out of time. Focus more on testing (e.g. the dijkstra class is an important class to add tests) and adding functionality
- I tried to run the BaseSimulation, but nothing happened and the corresponding test failed. Is this intentional?

File: src/simulation/locomotionAlgorithms/dijkstra.py (hunk 1 above)
Comment (AdiJoost) [reply src/simulation/locomotionAlgorithms/dijkstra.py:47]: Yes we are. The Step is not diagonal, one of the indices (x or y) is the same number as the new indices(xNew, yNew) therefore either x -xNew or y - yNew is zero.

File: src/dtoModule/tileValueDTO.py (hunk 6 above)
Comment (AdiJoost) [reply src/dtoModule/tileValueDTO.py:3]: DataTransfereObject. We use this base classes like Pojos in java.

File: src/simulation/locomotionAlgorithms/dijkstra.py (hunk 1 above)
Comment (DrGeli) [reply src/simulation/locomotionAlgorithms/dijkstra.py:47]: The method checks if either the substraction of the x coordintes are zero or the substraction of y coordintes. If one of each is true, it is NOT diagonal. Imho, the method should be named differently or should test that neither the  x substraction nor y substraction are allowed to be zero.

Comment (AdiJoost) [reply src/simulation/locomotionAlgorithms/dijkstra.py:47]: I'm sorry. Well spotted. Thank you.

**Pull request merged**